### PR TITLE
Refactor cluster autoscaler flags to be injected on-demand and not on-import

### PR DESCRIPTION
**Pull request by mtrqq**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This patch does following things:

* Regenerate FAQ.md with the newest set of flags, since it was updated some time ago and to prove that change doesn't have impact on the flags defined;
* Move flags initialization from global space to a an on-demand function to produce the autoscaling options;
* Add tests which verify that everything parses correctly and tests for edge cases in argument parsing;

This direction is on the same line as making cluster autoscaler a library in the long-run, current approach makes it infeasible for forks to properly extend and control their CLI arguments, with this patch - it's now possible without loosing anything else.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster-autoscaler/config/flags/flags.go [hunk 1]
```diff
@@ -55,410 +57,294 @@ func (flag *MultiStringFlag) Set(value string) error {
...
-	value := new(MultiStringFlag)
-	flag.Var(value, name, usage)
-	return value
+func (flag *MultiStringFlag) Type() string {
+	return "MultiStringFlag"
+}
+
+// AutoscalingFlags holds the command line flags for the autoscaler and responsible
+// for parsing and building the AutoscalingOptions struct from commandline flags.
+type AutoscalingFlags struct {
```
Comment (Choraden) [inline cluster-autoscaler/config/flags/flags.go:51]: The `AutoscalingFlags` struct acts as a hybrid: some CLI flags are bound directly to fields of the nested `p.o` (`config.AutoscalingOptions`), while other flags are bound to temporary fields for post-processing. This leads to a "half-baked" state.
What do you think about reworking `Options()` call so that it actually builds AutoscalingOptions from the raw fields?

File: cluster-autoscaler/config/flags/flags.go [hunk 2]
```diff
@@ -55,410 +57,294 @@ func (flag *MultiStringFlag) Set(value string) error {
...
+		p.o.MaxStartupTime = p.o.MaxInactivityTime
 	}
+
+	if p.o.MaxStartupTime < p.o.MaxFailingTime {
+		p.o.MaxStartupTime = p.o.MaxFailingTime
+	}
+
+	p.o.AllowedSchedulers = allowedSchedulers
+	p.o.BypassedSchedulers = scheduler_util.SchedulersMap(p.bypassedSchedulers)
+	p.o.DynamicResourceAllocationEnabled = p.o.DynamicResourceAllocationEnabled
```
Comment (Choraden) [inline cluster-autoscaler/config/flags/flags.go]: I believe this is a mistake

File: cluster-autoscaler/config/flags/flags.go [hunk 3]
```diff
@@ -55,410 +57,294 @@ func (flag *MultiStringFlag) Set(value string) error {
 	return nil
 }
 
-func multiStringFlag(name string, usage string) *MultiStringFlag {
-	value := new(MultiStringFlag)
-	flag.Var(value, name, usage)
-	return value
+func (flag *MultiStringFlag) Type() string {
```
Comment (Choraden) [inline cluster-autoscaler/config/flags/flags.go]: I guess, we are good to start using `fs.StringArrayVar` or `fs.StringSliceVar`

File: cluster-autoscaler/config/flags/flags_test.go [hunk 4]
```diff
@@ -194,13 +194,342 @@ func TestCreateAutoscalingOptions(t *testing.T) {
...
+			gotOptions, err := f.Options()
+			if err != nil {
+				t.Fatalf("Options() got unexpected error: %v", err)
+			}
 			tc.wantOptionsAsserter(t, gotOptions)
 		})
 	}
 }
+
+func TestAutoscalingFlagsAllPossible(t *testing.T) {
```
Comment (Choraden) [inline cluster-autoscaler/config/flags/flags_test.go:213]: I'm afraid, this will easily get out-dated. Do you see any protection against such case?

**Review by Choraden — COMMENTED** [review]
Thanks @mtrqq! Refactoring flags to an object will make them more flexible, context aware, and it will give more control for their users.

Left some comments.

File: cluster-autoscaler/config/flags/flags.go (hunk 3 above)
Comment (mtrqq) [reply cluster-autoscaler/config/flags/flags.go]: Good call, StringArrayVar has almost same implementation as we have for MultiStringFlag, replaced it!

File: cluster-autoscaler/config/flags/flags.go (hunk 2 above)
Comment (mtrqq) [reply cluster-autoscaler/config/flags/flags.go]: Thanks for noticing, removed the line

File: cluster-autoscaler/config/flags/flags.go (hunk 1 above)
Comment (mtrqq) [reply cluster-autoscaler/config/flags/flags.go:51]: I fully the concern, the alternative is either copying all the fields from AutoscalingOptions and duplicating them for the parsing purposes or introducing a flag parsing layer which will automatically translate strings into correct types. I see value in it, but I don't see an easy path to implement that for current edge cases, would that be find to implement this as a further improvement?

In the current PR I wanted just to move flags to a function invocation so that it will become a proper library interface and not the application one and not to rewrite the parsing logic on its own

File: cluster-autoscaler/config/flags/flags_test.go (hunk 4 above)
Comment (mtrqq) [reply cluster-autoscaler/config/flags/flags_test.go:213]: I agree that it will become stale at some point unless contributors would proactively extend it. But I think it should be a matter of a review process and policies and not the automated checks

File: cluster-autoscaler/config/flags/flags.go (hunk 1 above)
Comment (Choraden) [reply cluster-autoscaler/config/flags/flags.go:51]: Totally agree. Let's just move here.

File: cluster-autoscaler/config/flags/flags.go [hunk 5]
```diff
@@ -37,434 +36,305 @@ import (
...
-		NodeDeletionCandidateTTL:                     *nodeDeletionCandidateTTL,
-		CapacitybufferControllerEnabled:              *capacitybufferControllerEnabled,
-		CapacitybufferPodInjectionEnabled:            *capacitybufferPodInjectionEnabled,
-		CapacityBufferPodDryRunEnabled:               *capacityBufferPodDryRunEnabled,
-		NodeRemovalLatencyTrackingEnabled:            *nodeRemovalLatencyTrackingEnabled,
-		MaxNodeSkipEvalTimeTrackerEnabled:            *maxNodeSkipEvalTimeTrackerEnabled,
-		CapacityQuotasEnabled:                        *capacityQuotasEnabled,
-		ScaleUpSimulationForSkippedNodeGroupsEnabled: *scaleUpSimulationForSkippedNodeGroupsEnabled,
+	if p.o.MaxStartupTime < p.o.MaxInactivityTime {
+		p.o.MaxStartupTime = p.o.MaxInactivityTime
```
Comment (BigDarkClown) [inline cluster-autoscaler/config/flags/flags.go:311]: Missing warning

File: cluster-autoscaler/config/flags/flags.go [hunk 6]
```diff
@@ -37,434 +36,305 @@ import (
...
 	}
+
+	if p.o.MaxStartupTime < p.o.MaxFailingTime {
+		p.o.MaxStartupTime = p.o.MaxFailingTime
```
Comment (BigDarkClown) [inline cluster-autoscaler/config/flags/flags.go:315]: Missing warning

File: cluster-autoscaler/config/flags/flags.go [hunk 7]
```diff
@@ -37,434 +36,305 @@ import (
...
+	if p.schedulerConfigFile != "" {
+		parsedSchedConfig, err = scheduler_util.ConfigFromPath(p.schedulerConfigFile)
+		if err != nil {
+			return config.AutoscalingOptions{}, fmt.Errorf("Failed to get scheduler config: %w", err)
+		}
 	}
-	if err != nil {
-		klog.Fatalf("Failed to get scheduler config: %v", err)
+
+	if len(p.balancingLabels) > 0 && len(p.balancingIgnoreLabels) > 0 {
```
Comment (BigDarkClown) [inline cluster-autoscaler/config/flags/flags.go:273]: Why was this check added?

File: cluster-autoscaler/config/flags/flags.go [hunk 8]
```diff
@@ -37,434 +36,305 @@ import (
...
+		return config.AutoscalingOptions{}, fmt.Errorf("--enable-dynamic-resource-allocation flag must be true: %t", p.o.DynamicResourceAllocationEnabled)
 	}
 
-	if *scaleDownEnabled == false {
+	allowedSchedulers, err := parseAllowedSchedulers(p.allowedSchedulers, p.bypassedSchedulers)
+	if err != nil {
+		return config.AutoscalingOptions{}, err
+	}
+
+	if p.clusterSnapshotParallelism != ClusterSnapshotParallelismDefault {
```
Comment (BigDarkClown) [inline cluster-autoscaler/config/flags/flags.go:302]: Why was this override added?

**Review by BigDarkClown — COMMENTED** [review]
Overall LGTM. Found some missing warnings + I am not sure about one override and one check. I would keep this change as minimally changing as possible, PTAL.

File: cluster-autoscaler/config/flags/flags.go (hunk 5 above)
Comment (mtrqq) [reply cluster-autoscaler/config/flags/flags.go:311]: Added it back and stick to the previous defaulting logic, ty for noticing!

File: cluster-autoscaler/config/flags/flags.go (hunk 6 above)
Comment (mtrqq) [reply cluster-autoscaler/config/flags/flags.go:315]: Added it back and stick to the previous defaulting logic, ty for noticing!

File: cluster-autoscaler/config/flags/flags.go (hunk 7 above)
Comment (mtrqq) [reply cluster-autoscaler/config/flags/flags.go:273]: My oversight, likely confused it with a different flag

File: cluster-autoscaler/config/flags/flags.go (hunk 8 above)
Comment (mtrqq) [reply cluster-autoscaler/config/flags/flags.go:302]: I don't see any reason for it, replaced with straight assignment

**Comment by mtrqq** [comment]
PR got moved to new the repository! 

Link: https://github.com/kubernetes-sigs/cluster-autoscaler/pull/25
